analyzers/src/SonarAnalyzer.Common/Rules/SecurityPInvokeMethodShouldNotBeCalledBase.cs

```diff
@@ -0,0 +1,87 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class SecurityPInvokeMethodShouldNotBeCalledBase<TSyntaxKind, TInvocationExpressionSyntax> : SonarDiagnosticAnalyzer
+        where TSyntaxKind : struct
+        where TInvocationExpressionSyntax : SyntaxNode
+    {
+        protected const string DiagnosticId = "S3884";
+        protected const string MessageFormat = "Refactor the code to remove this use of '{0}'.";
+        protected const string InteropName = "ole32";
+        protected const string InteropDllName = InteropName + ".dll";
+
+        protected abstract TSyntaxKind SyntaxKind { get; }
+        protected abstract ILanguageFacade Language { get; }
+
+        protected abstract SyntaxNode Expression(TInvocationExpressionSyntax invocationExpression);
+        protected abstract SyntaxToken Identifier(SyntaxNode syntaxNode);
+        protected abstract IMethodSymbol MethodSymbolForInvalidInvocation(SyntaxNode syntaxNode, SemanticModel semanticModel);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected DiagnosticDescriptor Rule { get; }
+
+        protected ISet<string> InvalidMethods { get; } = new HashSet<string>
+        {
+            "CoSetProxyBlanket",
+            "CoInitializeSecurity"
+        };
+
+        protected SecurityPInvokeMethodShouldNotBeCalledBase(System.Resources.ResourceManager rspecResources) =>
+            Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, rspecResources);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer, CheckForIssue, SyntaxKind);
+
+        protected virtual bool IsImportFromInteropDll(ISymbol symbol) =>
```

Comment:
For the coverage of `SecurityPInvokeMethodShouldNotBeCalledBase.cs` , [lines 63-66](https://sonarcloud.io/code?id=sonaranalyzer-dotnet&pullRequest=3997&selected=sonaranalyzer-dotnet%3Aanalyzers%2Fsrc%2FSonarAnalyzer.Common%2FRules%2FSecurityPInvokeMethodShouldNotBeCalledBase.cs&line=63), there is room for coverage improvement.

`DllImportAttribute` has [several fields](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.dllimportattribute.bestfitmapping?view=net-5.0) which can be passed to the Attribute creation.

You could add some snippets that compile with different combinations of arguments, like:

```
    <DllImport("oLe32" , BestFitMapping:=false, CallingConvention:=CallingConvention.FastCall)>
    Public Shared Function CoSetProxyBlanket22() As Integer
    End Function
```

And some snippets with compile errors (that can happen in the IDE, as the analysis is done during build on WIP code)

```
        [DllImport(BestFitMapping = false)] // Error [CS7036]
        public static extern int CoInitializeSecurity1();

        [DllImport(BestFitMapping =false, EntryPoint="ole32.dll", ExactSpelling = true)] // Error [CS7036]
        public static extern int CoInitializeSecurity2();
```